src/Control/SessionHandler/DatabaseSessionHandler.php

```diff
@@ -0,0 +1,230 @@
+     */
+    public function read(#[SensitiveParameter] string $id): string|false
+    {
+        if (!$this->isDatabaseReady()) {
+            $this->logError('Could not read session - database is not ready');
+            return false;
+        }
+
+        /** @var Query $rows */
+        $rows = DB::withPrimary(fn() => $this->getSqlSelect($id)->execute());
+        if ($rows->numRecords() === 0) {
+            return '';
+        }
+        return $rows->record()['Data'];
+    }
+
+    /**
+     * @inheritDoc
+     * Writes session data to a cache entry.
+     */
+    public function write(#[SensitiveParameter] string $id, string $data): bool
+    {
+        if (!$this->isDatabaseReady()) {
+            $this->logError('Could not write session - database is not ready');
+            return false;
+        }
+
+        $table = Convert::symbol2sql(static::config()->get('table_name'));
+        $where = [Convert::symbol2sql('ID') => $id];
+        if ($this->sessionExists($id, true)) {
+            $query = SQLUpdate::create($table, where: $where);
+        } else {
+            $query = SQLInsert::create($table, $where);
+        }
+        $query->addAssignments([
+            Convert::symbol2sql('Data') => $data,
+            Convert::symbol2sql('Expiry') => DBDatetime::now()->getTimestamp() + $this->getLifetime(),
+        ]);
+        $query->execute();
+        return true;
+    }
+
+    /**
+     * @inheritDoc
+     * A session ID is valid if an entry for that session ID already exists and has not expired.
+     */
+    public function validateId(#[SensitiveParameter] string $id): bool
+    {
+        if (!$this->isDatabaseReady()) {
+            $this->logError('Could not validate session ID - database is not ready');
+            return false;
+        }
+        return $this->sessionExists($id);
+    }
+
+    /**
+     * @inheritDoc
+     * Called instead of write if session.lazy_write is enabled and no data has changed for this session.
+     */
+    public function updateTimestamp(#[SensitiveParameter] string $id, string $data): bool
+    {
+        // The logic for updating the timestamp ends up being effectively identical to just
+        // writing the session - there's no optimisation to be made by using separate logic.
+        return $this->write($id, $data);
+    }
+
+    /**
+     * Add the database table. This is called by an extension when building the db.
+     * Note that we don't just use a DataObject because:
+     * 1. We don't want things like versioning, fluent, etc to ever be able to affect sessions
+     * 2. We don't want developers to be affecting db operations via hooks (interact with sessions with the Session class)
+     * 3. We don't want sessions to be used in any other ways that DataObjects are often
+     * 4. We only want to build the table if this is the configured save handler
+     */
+    public function requireTable()
```

Comment:
```suggestion
    public function requireTable(): void
```
Covariance allows more specific return types

Reply:
Done